### PR TITLE
fix: test_chunk_forwarding_optimization

### PR DESCRIPTION
**Pull request by jancionear**
This test didn't work with stateless validation. Let's fix it by properly handling `ChunkStateWitness` and `ChunkEndorsement` messages.

Refs: https://github.com/near/nearcore/issues/10506

### CODE REVIEW

File: integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs [hunk 1]
```diff
@@ -603,7 +601,11 @@ impl ChunkForwardingOptimizationTestData {
         }
     }
 
-    fn process_one_peer_message(&mut self, client_id: usize, requests: NetworkRequests) {
+    fn process_one_peer_message(
+        &mut self,
```
Comment (shreyan-gupta) [inline integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs:605]: Why is the mut necessary here?

File: integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs [hunk 2]
```diff
@@ -680,30 +684,27 @@ impl ChunkForwardingOptimizationTestData {
...
+                let network_adapter = self.env.network_adapters[i].clone();
+
+                any_message_processed |= network_adapter.handle_filtered(|request| match request {
+                    PeerManagerMessageRequest::NetworkRequests(requests) => {
+                        self.process_one_peer_message(i, requests).map(|ignored_request| {
+                            PeerManagerMessageRequest::NetworkRequests(ignored_request)
+                        })
                     }
-                }
+                    _ => None,
```
Comment (shreyan-gupta) [inline integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs:706]: Shouldn't this be `_ => Some(request)`?

File: integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs [hunk 3]
```diff
@@ -770,6 +763,12 @@ fn test_chunk_forwarding_optimization() {
             assert_eq!(&accepted_blocks[0], block.header().hash());
             assert_eq!(test.env.clients[i].chain.head().unwrap().height, block.header().height());
         }
+
+        test.process_network_messages();
+        test.env.process_shards_manager_responses(0);
+
+        let output = test.env.propagate_chunk_state_witnesses();
```
Comment (shreyan-gupta) [inline integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs:770]: nit: Can add a tiny comment here saying something on the lines of, we're not really testing state witness or chunk endorsement but this is necessary for block production.

**Review by shreyan-gupta — APPROVED** [review]

File: integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs (hunk 2 above)
Comment (jancionear) [reply integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs]: Right, it should be `Some`. Fixed :+1:
The original code panicked in this case, but we can just ignore non-network messages, it doesn't matter.

File: integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs (hunk 3 above)
Comment (jancionear) [reply integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs:771]: Added a comment

File: integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs (hunk 1 above)
Comment (jancionear) [reply integration-tests/src/tests/client/features/access_key_nonce_for_implicit_accounts.rs:605]: When a message is received the test environment processes it, which requires a mutable reference to `self`.
For example when a `PartialEncodedChunkRequest` is received, we have to call:
```rust
self.env.process_partial_encoded_chunk_request(
    client_id,
    PeerManagerMessageRequest::NetworkRequests(requests),
);
```

**Comment by jancionear** [comment]
Okay, maybe there's more to it.
I rebased onto the latest `master` and the test started failing on my local machine as well.
It looks like the latest changes broke something, I'll need to take a deeper look.

**Comment by jancionear** [comment]
It looks like #10531 broke the test, after #10531 block producers no longer send chunk endorsements to themselves, and because of that `wait_to_propagate_chunk_endorsements` timed out. In the log I can see that the missing endorsements are all from `test0`, who is the only block producer in this test:
```
Missing chunk endorsements: {ChunkHash(CFvKbwTVCSMXnduvFta5cXXRScSRnAyRvS3Jh7YnJM5j): {AccountId("test0")}, ChunkHash(8SWhRovasBrhXvHBVstgwNnffv72PhJJNxUjbN3XPHjs): {AccountId("test0")}, ChunkHash(DwLwocjKBBS6QhzBAk4axm5xDzDw9XHG6qq5goZ2VwXM): {AccountId("test0")}, ChunkHash(Bpmoe5YLnsK7B7wxkCMTiEQiPxdYzMRQtpsVK9DEdw6A): {AccountId("test0")}}
```

I fixed it by adding an `excluded_block_producer` argument to `wait_to_propagate_chunk_endorsements`. We will wait for endorsements from everyone except the next block producer, as this block producer won't sent an endorsement to themselves. This makes the test pass again.

An interesting question is - why didn't it break `run_chunk_validation_test()`? I think the reason is that in this test there are multiple block producers, and Client sends endorsements to the next N block producers, so the next block producer will send out the endorsement to the next next block producer, thanks to which `wait_to_propagate_chunk_endorsements` doesn't time out.

**Comment by shreyan-gupta** [comment]
Yeah, seems like we've identified a gap in our test framework here. We send out the endorsements to the next 5 block producers, which means `wait_to_propagate_chunk_endorsements` should ideally be checking for 5 copies of an endorsement :(

**Comment by shreyan-gupta** [comment]
I don't like the solution of passing the block producer but I really couldn't find a better way of doing this. So let's merge this fix in for now.

**Review by shreyan-gupta — APPROVED** [review]

**Pull request merged**